### PR TITLE
Sequent: Fix exRight and allLeft

### LINKED ISSUE #104: FO Sequent: Error when using allLeft or exRight
When using allLeft or exRight in the FO sequent calculus and instantiating with a term that is not a simple constant, e.g., `f(a)`, the following error is shown by the UI:
```
Character at position 1 in 'f(a)' is not allowed in constants
```

How to reproduce:

1. Enter `\all X: P(X) -> P(f(a))`
2. Use impRight
3. Use allLeft with `f(a)`
4. Get the error

### PR DESCRIPTION
**Pull request by Drodt**
Fixes #104 

Changes:
- Backend:
  - AllLeft, AllRight, ExLeft, and ExRight no longer take parameter `varAssign` of type `Map<String, String>` but `instTerm` of type `String`. This is a simplification, as the map contained only one element anyway and that must be an assignment for the bound variable by the respective quantifier.
  - `instTerm` is now allowed to be _any_ String that can be parsed to a term---i.e., not just a constant---for AllLeft and ExRight
  - Add `Signature` class representing the signature of a formula and `SignatureExtractor`
  - Replace `IdentifierCollector` with `SignatureExtractor` (=> we only need to search the names of, e.g., functions)
  - Check arity of functions/relations during parsing
    - Prevent using the same identifier for a function and constant / a relation and bound var
  - Check arity when instantiating a term
  - Implement automatic constant finding for AllRight/ExLeft
  - Tests for the new functionality
- Frontend:
  - Updated many dependencies
  - Bumped node version to 18
  - No longer rely on `preact-cli` as it is outdated
  - Now build with [vite](https://vitejs.dev/) 
  - Transformed the scss (style) files to scss modules (a new standard)
  - Switch from TSLint to ESLint, as TSLint is no longer suppoerted
  - Change frontend according to backend API changes

TODO:
- [x] Arity checks for other calculi
- [x] Add automatic button to frontend for AllRight/ExLeft
- [x] Tests for Signature stuff

### CODE REVIEW

**Comment by Drodt** [comment]
This PR has many breaking changes. We should carefully review the changes.

File: backend/src/main/kotlin/kalkulierbar/logic/transform/Signature.kt [hunk 1]
```diff
@@ -0,0 +1,230 @@
+package main.kotlin.kalkulierbar.logic.transform
```
Comment (rec0de) [inline backend/src/main/kotlin/kalkulierbar/logic/transform/Signature.kt:1]: Package should be kalkulierbar.logic.transform for consistenty

File: backend/src/main/kotlin/kalkulierbar/sequent/fosc/BasicMoves.kt [hunk 2]
```diff
@@ -223,43 +193,45 @@ fun applyExRight(state: FOSCState, nodeID: Int, listIndex: Int, varAssign: Map<S
...
-        throw IllegalMove("Constant '$str' does not start with a lowercase letter")
-
-    for (i in str.indices) {
-        if (!Tokenizer.isAllowedChar(str[i]))
-            throw IllegalMove("Character at position $i in '$str' is not allowed in constants")
+    val lower = quantVar.lowercase()
+    var newName = lower + "_0"
+
+    while (sig.hasConstOrFunction(newName)) {
+        val m = REGEX_NUMERIC_SUFFIX.find(newName)!!
```
Comment (rec0de) [inline backend/src/main/kotlin/kalkulierbar/sequent/fosc/BasicMoves.kt:232]: Is there a reason we need the regex?

Why not, say

```
val baseName = quantVar.lower()
var idx = 0
while(sig.hasConstOrFunction("$baseName_$idx")) {
    i += 1
}
```

File: backend/src/main/kotlin/kalkulierbar/sequent/fosc/BasicMoves.kt [hunk 3]
```diff
@@ -223,43 +193,45 @@ fun applyExRight(state: FOSCState, nodeID: Int, listIndex: Int, varAssign: Map<S
...
 
+val REGEX_NUMERIC_SUFFIX = Regex("^(.+)_(\\d+)$")
+
 /**
  * Tries to find a variable Name which leads to solving the proof
  */
-private fun findFittingVariableName(node: TreeNode): String {
-    throw IllegalMove("Not yet implemented")
-}
+private fun findFittingVariableName(node: TreeNode, quantVar: String): String {
```
Comment (rec0de) [inline backend/src/main/kotlin/kalkulierbar/sequent/fosc/BasicMoves.kt:225]: I feel like naming this `findFreshVariable` (or constant? these are constants right?) would be more accurate, esp considering possible future functions searching actual fitting assignment in a proof-search sense

**Review by rec0de — COMMENTED** [review]
Backend changes looking good :) Left some comments for minor consistency / naming questions

File: backend/src/main/kotlin/kalkulierbar/logic/transform/Signature.kt (hunk 1 above)
Comment (Drodt) [reply backend/src/main/kotlin/kalkulierbar/logic/transform/Signature.kt:1]: Don't you love that both are possible and IDEs disagree about that? ^^

Done

**Comment by Drodt** [comment]
While we're at it: Should we rename the sequent files/folders to be more consistent with the other calculi? @rec0de @JuliusHenke 

if so: Should I open a separate PR for this?

**Comment by rec0de** [comment]
If you want to, sure - no strong feelings either way

**Comment by JuliusHenke** [comment]
> While we're at it: Should we rename the sequent files/folders to be more consistent with the other calculi?

Consistent naming is always nice 👍 Thanks for the great PR and all of the tooling changes.

**Pull request merged**